phd-tests/framework/src/test_vm/metrics.rs

```diff
@@ -0,0 +1,200 @@
+            OximeterProducerInfo { addr: info.address, uuid: info.id };
+
+        // There should always be at least one Receiver on the channel since we
+        // hold one in `self`.
+        self.sampler_sender
+            .send(Some(new_sampler))
+            .expect("channel is subscribed");
+    }
+}
+
+impl FakeOximeterSampler {
+    /// Sample Propolis' Oximeter metrics, taking some function that determines
+    /// if a sample is satisfactory for the caller to proceed with.
+    ///
+    /// `wait_for_propolis_stats` will poll the corresponding Oximeter producer
+    /// and call `f` with each returned set of results.
+    ///
+    /// Panics if `f` does not return `Some` after some number of retries and
+    /// `ProducerResults` updates.
+    pub async fn wait_for_propolis_stats<U>(
+        &self,
+        f: impl Fn(ProducerResults) -> Option<U>,
+    ) -> U {
+        let result = backoff::future::retry(
+            backoff::ExponentialBackoff {
+                max_interval: Duration::from_secs(1),
+                max_elapsed_time: Some(Duration::from_secs(10)),
+                ..Default::default()
+            },
+            || async {
+                let producer_results = self.sample_propolis_stats().await
+                    .map_err(backoff::Error::transient)?;
+
+                if let Some(metrics) = f(producer_results) {
+                    Ok(metrics)
+                } else {
+                    Err(backoff::Error::transient(anyhow::anyhow!(
+                        "full metrics sample not available or fresh enough (yet?)"
+                    )))
+                }
+            },
+        )
+        .await;
+
+        result.expect("propolis-server Oximeter stats should become available")
+    }
+
+    /// Sample Propolis' Oximeter metrics, including the timestamp of the oldest
+    /// metric reflected in the sample.
+    ///
+    /// Returns `None` for some kinds of incomplete stats or when no stats are
+    /// available at all.
+    async fn sample_propolis_stats(
+        &self,
+    ) -> Result<ProducerResults, anyhow::Error> {
+        let metrics_url = {
+            self.sampler
+                .clone()
+                .wait_for(Option::is_some)
+                .await
+                .expect("can recv");
+            let sampler = self.sampler.borrow();
+            let stats = sampler.as_ref().expect("sampler does not become None");
+            format!("http://{}/{}", stats.addr, stats.uuid)
+        };
+        let res = reqwest::Client::new()
+            .get(metrics_url)
+            .send()
+            .await
+            .expect("can send oximeter stats request");
+        assert!(
+            res.status().is_success(),
+            "failed to fetch stats from propolis-server"
+        );
+        trace!(?res, "got stats response");
+        Ok(res.json::<ProducerResults>().await?)
+    }
+}
+
+// Stub functionality for our fake Nexus that test Oximeter produces
```

Comment:
```suggestion
// Stub functionality for our fake Nexus that test Oximeter producers
```

Reply:
(included in e88d07c638c9ef3c103ba9f34f41b44c1c458807, ty)